File: packages/weather/src/mostlyright/weather/_fetchers/_iem_mos.py
"""IEM MOS (Model Output Statistics) fetcher — JSON from mesonet.agron.iastate.edu.

The parity-compatible forecast source for ``research(include_forecast=True)``
Mode 1. Output matches ``schema.forecast.iem_mos.v1``.

Per-model coverage: NBE (preferred for US), GFS, LAV, MET, ECM.
Endpoint: ``https://mesonet.agron.iastate.edu/api/1/mos.json``.
Polite floor: 1 req/sec per IEM CLAUDE.md guideline (reused from iem_asos).
"""

from __future__ import annotations

import logging
import time
from datetime import UTC, datetime, timedelta
from typing import Any

import httpx
import pandas as pd
from mostlyright._internal._http import HTTP_TIMEOUT

log = logging.getLogger(__name__)

_IEM_MOS_URL = "https://mesonet.agron.iastate.edu/api/1/mos.json"

#: MOS models IEM exposes through the JSON endpoint.
SUPPORTED_MOS_MODELS: frozenset[str] = frozenset({"nbe", "gfs", "lav", "met", "ecm"})

#: Conversion factor from knots to metres per second (canonical column unit
#: is ``wind_speed_ms`` per :mod:`mostlyright.core.schemas.forecast`).
_KT_TO_MS: float = 0.5144444

#: IEM polite-floor delay between MOS GETs (reused convention from iem_asos).
_MOS_POLITE_DELAY_S: float = 1.0

#: NBE runtime-cycle transition: IEM moved NBE from {01,07,13,19}Z to
#: {00,06,12,18}Z on 2026-05-05. Historical backfills that span this
#: cutover need both sets covered.
_NBE_CYCLE_CUTOVER: datetime = datetime(2026, 5, 5, tzinfo=UTC)


def _runtime_hours_for(model: str, from_dt: datetime, to_dt: datetime) -> tuple[int, ...]:
    """Return the runtime-hours tuple for ``model`` in ``[from_dt, to_dt]``.

    NBE: ``{01,07,13,19}Z`` pre-cutover, ``{00,06,12,18}Z`` post-cutover.
    Ranges that span the cutover get the UNION so neither half silently
    returns empty 404s (Phase 17 Wave 3 iter-4 review).

    All other models default to ``{00,06,12,18}Z`` (GFS / LAV / MET / ECM
    follow the standard 6-hourly NCEP cycle).
    """
    if model == "nbe":
        pre = from_dt < _NBE_CYCLE_CUTOVER
        post = to_dt >= _NBE_CYCLE_CUTOVER
        if pre and post:
            return (0, 1, 6, 7, 12, 13, 18, 19)
        if pre:
            return (1, 7, 13, 19)
        return (0, 6, 12, 18)
    return (0, 6, 12, 18)


#: Canonical column list matching ``schema.forecast.iem_mos.v1.COLUMNS``.
#: Returned empty DataFrame uses these so dtype inference is consistent
#: across empty / populated paths.
_CANONICAL_COLUMNS: tuple[str, ...] = (
    "station",
    "issued_at",
    "valid_at",
    "forecast_hour",
    "model",
    "temp_c",
    "dew_point_c",
    "wind_speed_ms",
    "wind_dir_deg",
    "precip_probability",
    "sky_cover_pct",
    "source",
    "retrieved_at",
)


def _f_to_c(temp_f: float | None) -> float | None:
    """Fahrenheit → Celsius, ``None`` passthrough."""
    if temp_f is None:
        return None
    return (float(temp_f) - 32.0) * 5.0 / 9.0


def _kt_to_ms(kt: float | None) -> float | None:
    """Knots → m/s, ``None`` passthrough."""
    if kt is None:
        return None
    return float(kt) * _KT_TO_MS


def _pct_to_unit(pct: float | None) -> float | None:
    """Percent (0-100) → probability (0-1), ``None`` passthrough."""
    if pct is None:
        return None
    return float(pct) / 100.0


def _maybe_float(row: dict[str, Any], key: str) -> float | None:
    """Pull a numeric field from an IEM MOS row, treating ``M`` / blank /
    null as missing.
    """
    v = row.get(key)
    if v is None or v == "" or v == "M":
        return None
    try:
        return float(v)
    except (TypeError, ValueError):
        return None


def _parse_dt(value: str | None) -> datetime | None:
    """Parse an ISO datetime from IEM MOS, returning UTC-aware or ``None``."""
    if not value:
        return None
    try:
        # IEM sometimes returns trailing 'Z' — normalise to UTC offset.
        normalized = value.replace("Z", "+00:00")
        dt = datetime.fromisoformat(normalized)
    except (TypeError, ValueError):
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=UTC)
    return dt.astimezone(UTC)


def _parse_mos_row(
    row: dict[str, Any],
    *,
    station: str,
    model: str,
    retrieved_at: datetime,
) -> dict[str, Any] | None:
    """Project one IEM MOS row to the canonical schema. Returns ``None``
    when both ``runtime`` and ``ftime`` are unparseable (the row is
    structurally invalid).
    """
    issued_at = _parse_dt(row.get("runtime") or row.get("model_runtime"))
    valid_at = _parse_dt(row.get("ftime") or row.get("valid_time"))
    if issued_at is None or valid_at is None:
        return None
    forecast_hour = round((valid_at - issued_at).total_seconds() / 3600)
    return {
        "station": station,
        "issued_at": issued_at,
        "valid_at": valid_at,
        "forecast_hour": forecast_hour,
        "model": model.upper(),
        "temp_c": _f_to_c(_maybe_float(row, "tmp")),
        "dew_point_c": _f_to_c(_maybe_float(row, "dpt")),
        "wind_speed_ms": _kt_to_ms(_maybe_float(row, "wsp")),
        "wind_dir_deg": (round(d) if (d := _maybe_float(row, "wdr")) is not None else None),
        "precip_probability": _pct_to_unit(_maybe_float(row, "pop12")),
        "sky_cover_pct": None,  # IEM MOS doesn't expose a single sky-cover %
        "source": "iem.archive",
        "retrieved_at": retrieved_at,
    }


def fetch_iem_mos(
    station: str,
    from_date: str,
    to_date: str,
    *,
    model: str = "nbe",
    client: httpx.Client | None = None,
    timeout: float = HTTP_TIMEOUT,
) -> pd.DataFrame:
    """Fetch IEM MOS forecasts for ``station`` + ``model`` in
    ``[from_date, to_date]`` (inclusive, ISO ``YYYY-MM-DD``).

    Iterates the 00 / 06 / 12 / 18 Z runtimes in the requested date
    range, GETs each as ``runtime`` query parameter, and accumulates the
    returned per-ftime rows into a single DataFrame matching
    ``schema.forecast.iem_mos.v1``.

    Args:
        station: ICAO code (e.g. ``"KNYC"``).
        from_date: ISO ``YYYY-MM-DD`` lower bound (inclusive).
        to_date: ISO ``YYYY-MM-DD`` upper bound (inclusive).
        model: One of :data:`SUPPORTED_MOS_MODELS`. Default ``"nbe"``.
        client: Optional :class:`httpx.Client` to reuse.
        timeout: Per-request timeout in seconds.

    Returns:
        DataFrame with the canonical columns. Empty (with columns) if
        IEM returns no MOS rows in the requested window.

    Raises:
        ValueError: ``model`` not in :data:`SUPPORTED_MOS_MODELS` or
            date strings malformed.
        httpx.HTTPError: IEM unreachable / non-2xx (404 is silently
            skipped — many runtimes have no data and that's normal).
    """
    if model not in SUPPORTED_MOS_MODELS:
        raise ValueError(f"model must be one of {sorted(SUPPORTED_MOS_MODELS)}; got {model!r}")
    try:
        from_dt = datetime.fromisoformat(from_date).replace(tzinfo=UTC)
        to_dt = datetime.fromisoformat(to_date).replace(tzinfo=UTC, hour=23, minute=59, second=59)
    except ValueError as exc:
        raise ValueError(
            f"from_date / to_date must be ISO YYYY-MM-DD; "
            f"got from_date={from_date!r} to_date={to_date!r}"
        ) from exc

    close_client = client is None
    if client is None:
        client = httpx.Client(timeout=timeout)

    rows: list[dict[str, Any]] = []
    retrieved_at = datetime.now(UTC)
    try:
        # Phase 17 Wave 3 iter-4 review: model-aware runtime hours.
        # NBE cycle hours moved from {01,07,13,19}Z to {00,06,12,18}Z on
        # 2026-05-05; pre-cutover historical backfills would 404 on every
        # canonical-hour request and silently return an empty DataFrame.
        # The safe strategy is to iterate BOTH hour sets when the cycle
        # transition is in or near the requested window. The 404-skip
        # path absorbs the wrong-runtime requests at the cost of ~1
        # extra GET per day per side.
        runtimes: list[datetime] = []
        hours = _runtime_hours_for(model, from_dt, to_dt)
        cur = from_dt.replace(hour=0, minute=0, second=0, microsecond=0)
        while cur <= to_dt:
            for h in hours:
                rt = cur.replace(hour=h)
                if from_dt <= rt <= to_dt:
                    runtimes.append(rt)
            cur = cur + timedelta(days=1)

        for rt in runtimes:
            params = {
                "station": station,
                # IEM /api/1/mos.json regex ^(AVN|GFS|...|NBE|...)$ is
                # uppercase-only; sending lowercase returns HTTP 422
                # (issue #17). Mirrors the schema-column upper() at L153.
                "model": model.upper(),
                "runtime": rt.isoformat(),
            }
            resp = client.get(_IEM_MOS_URL, params=params)
            if resp.status_code == 404:
                log.debug(
                    "IEM MOS 404 for %s %s runtime=%s",
                    station,
                    model,
                    rt.isoformat(),
                )
                time.sleep(_MOS_POLITE_DELAY_S)
                continue
            resp.raise_for_status()
            payload = resp.json()
            for raw_row in payload.get("data", []):
                projected = _parse_mos_row(
                    raw_row,
                    station=station,
                    model=model,
                    retrieved_at=retrieved_at,
                )
                if projected is not None:
                    rows.append(projected)
            time.sleep(_MOS_POLITE_DELAY_S)
    finally:
        if close_client:
            client.close()

    if not rows:
        empty = pd.DataFrame(columns=list(_CANONICAL_COLUMNS))
        empty.attrs["source"] = "iem.archive"
        empty.attrs["retrieved_at"] = retrieved_at
        return _coerce_canonical_dtypes(empty)
    df = pd.DataFrame(rows, columns=list(_CANONICAL_COLUMNS))
    df.attrs["source"] = "iem.archive"
    df.attrs["retrieved_at"] = retrieved_at
    return _coerce_canonical_dtypes(df)


def _coerce_canonical_dtypes(df: pd.DataFrame) -> pd.DataFrame:
    """Coerce IEM MOS columns to the canonical
    ``schema.forecast.iem_mos.v1`` dtypes (Phase 17 Wave 3 iter-2 review).

    Uses pandas nullable dtypes (``Int64``, ``Float64``, ``string``) so
    missing fields survive without forcing the column to ``object``.
    Empty DataFrames also get the right per-column dtype so
    :func:`validate_dataframe` accepts them.
    """
    df["station"] = df["station"].astype("string")
    df["model"] = df["model"].astype("string")
    df["source"] = df["source"].astype("string")
    df["issued_at"] = pd.to_datetime(df["issued_at"], utc=True, errors="coerce")
    df["valid_at"] = pd.to_datetime(df["valid_at"], utc=True, errors="coerce")
    df["retrieved_at"] = pd.to_datetime(df["retrieved_at"], utc=True, errors="coerce")
    df["forecast_hour"] = pd.to_numeric(df["forecast_hour"], errors="coerce").astype("Int64")
    df["wind_dir_deg"] = pd.to_numeric(df["wind_dir_deg"], errors="coerce").astype("Int64")
    df["sky_cover_pct"] = pd.to_numeric(df["sky_cover_pct"], errors="coerce").astype("Int64")
    for fcol in ("temp_c", "dew_point_c", "wind_speed_ms", "precip_probability"):
        df[fcol] = pd.to_numeric(df[fcol], errors="coerce").astype("Float64")
    return df


__all__ = ["SUPPORTED_MOS_MODELS", "fetch_iem_mos"]

File: packages-ts/weather/src/forecasts/iem-mos.ts
// Phase 17 PLAN-11 — IEM MOS TS fetcher.
//
// Mirrors Python `packages/weather/src/mostlyright/weather/_fetchers/_iem_mos.py`.
// Endpoint: https://mesonet.agron.iastate.edu/api/1/mos.json
// CORS: OPEN per IEM ASOS posture (see `.planning/research/FORECAST-CORS-MATRIX.md`).

import type { IemMosModel, IemMosOptions, IemMosRow, IemMosSource } from "./types.js";

const IEM_MOS_URL = "https://mesonet.agron.iastate.edu/api/1/mos.json";

const SUPPORTED_MODELS: ReadonlySet<IemMosModel> = new Set(["nbe", "gfs", "lav", "met", "ecm"]);

const KT_TO_MS = 0.5144444;

const NBE_CYCLE_CUTOVER = Date.UTC(2026, 5 - 1, 5, 0, 0, 0); // 2026-05-05T00:00:00Z

/**
 * Max number of MOS runtime-cycle requests in flight at once (GH #58).
 *
 * The fan-out is bounded rather than unbounded: a `Promise.all` over the full
 * day × runtime-hour grid would launch ~1,460 simultaneous requests for a
 * year-scale window, risking client connection limits, memory pressure, and
 * IEM-side rate limiting. A cap of 8 keeps essentially all of the small-window
 * speedup (typical ±1–3 day windows have ≤ 12 cycles) while staying polite for
 * large historical ranges.
 */
export const MOS_FETCH_CONCURRENCY = 8;

/**
 * Map `items` through `fn` with at most `limit` invocations in flight at once,
 * returning results in input order (NOT resolution order). Bounded-concurrency
 * replacement for `Promise.all(items.map(fn))` — preserves the byte-identical
 * ordering the serial path produced while capping peak fan-out.
 */
async function mapWithConcurrency<T, R>(
  items: readonly T[],
  limit: number,
  fn: (item: T, index: number) => Promise<R>,
): Promise<R[]> {
  const results = new Array<R>(items.length);
  let cursor = 0;
  async function worker(): Promise<void> {
    while (cursor < items.length) {
      const index = cursor++;
      results[index] = await fn(items[index] as T, index);
    }
  }
  const poolSize = Math.min(limit, items.length);
  await Promise.all(Array.from({ length: poolSize }, () => worker()));
  return results;
}

/** Pick the right NBE runtime-hour set based on the requested range. */
function runtimeHoursFor(model: IemMosModel, fromDt: Date, toDt: Date): readonly number[] {
  if (model !== "nbe") return [0, 6, 12, 18];
  const fromMs = fromDt.getTime();
  const toMs = toDt.getTime();
  const pre = fromMs < NBE_CYCLE_CUTOVER;
  const post = toMs >= NBE_CYCLE_CUTOVER;
  if (pre && post) return [0, 1, 6, 7, 12, 13, 18, 19];
  if (pre) return [1, 7, 13, 19];
  return [0, 6, 12, 18];
}

function fahrenheitToCelsius(f: number | null): number | null {
  if (f === null || Number.isNaN(f)) return null;
  return ((f - 32) * 5) / 9;
}

function knotsToMs(kt: number | null): number | null {
  if (kt === null || Number.isNaN(kt)) return null;
  return kt * KT_TO_MS;
}

function percentToUnit(pct: number | null): number | null {
  if (pct === null || Number.isNaN(pct)) return null;
  return pct / 100;
}

function maybeNumber(value: unknown): number | null {
  if (value === null || value === undefined || value === "M" || value === "") {
    return null;
  }
  const num = typeof value === "number" ? value : Number(value);
  if (Number.isNaN(num)) return null;
  return num;
}

function parseDate(value: unknown): Date | null {
  if (typeof value !== "string" || !value) return null;
  const dt = new Date(value);
  if (Number.isNaN(dt.getTime())) return null;
  return dt;
}

interface RawMosRow {
  readonly runtime?: string;
  readonly model_runtime?: string;
  readonly ftime?: string;
  readonly valid_time?: string;
  readonly tmp?: number | string | null;
  readonly dpt?: number | string | null;
  readonly wsp?: number | string | null;
  readonly wdr?: number | string | null;
  readonly pop12?: number | string | null;
}

function parseRow(
  raw: RawMosRow,
  station: string,
  model: IemMosModel,
  retrievedAt: string,
): IemMosRow | null {
  const issuedDt = parseDate(raw.runtime ?? raw.model_runtime);
  const validDt = parseDate(raw.ftime ?? raw.valid_time);
  if (issuedDt === null || validDt === null) return null;
  const forecastHour = Math.round((validDt.getTime() - issuedDt.getTime()) / 3_600_000);
  return {
    station,
    model: model.toUpperCase(),
    issuedAt: issuedDt.toISOString(),
    validAt: validDt.toISOString(),
    forecastHour,
    tempC: fahrenheitToCelsius(maybeNumber(raw.tmp)),
    dewPointC: fahrenheitToCelsius(maybeNumber(raw.dpt)),
    windSpeedMs: knotsToMs(maybeNumber(raw.wsp)),
    windDirDeg: (() => {
      const d = maybeNumber(raw.wdr);
      return d === null ? null : Math.round(d);
    })(),
    precipProbability: percentToUnit(maybeNumber(raw.pop12)),
    skyCoverPct: null,
    source: "iem.archive" as IemMosSource,
    retrievedAt,
  };
}

/** Parse ISO `YYYY-MM-DD` to a UTC `Date` at 00:00:00. */
function parseIsoDate(iso: string, endOfDay: boolean): Date {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(iso);
  if (match === null) {
    throw new Error(`iemMosForecasts: from/to dates must be ISO YYYY-MM-DD; got ${iso}`);
  }
  const [, y, m, d] = match;
  if (endOfDay) {
    return new Date(Date.UTC(Number(y), Number(m) - 1, Number(d), 23, 59, 59));
  }
  return new Date(Date.UTC(Number(y), Number(m) - 1, Number(d)));
}

/**
 * Fetch IEM MOS forecasts for `station` in `[fromDate, toDate]`.
 *
 * Mirrors Python `fetch_iem_mos(...)`. Iterates the model's runtime-hour
 * grid (NBE moved from {01,07,13,19}Z to {00,06,12,18}Z on 2026-05-05;
 * other models use {00,06,12,18}Z), GETs the JSON endpoint, and projects
 * rows to {@link IemMosRow}.
 *
 * 404 responses are silently skipped (many runtimes have no MOS data).
 * Empty input range returns `[]`.
 *
 * @throws `Error` if `model` is not in `SUPPORTED_MODELS`.
 */
export async function iemMosForecasts(
  station: string,
  fromDate: string,
  toDate: string,
  opts: IemMosOptions = {},
): Promise<IemMosRow[]> {
  const model = opts.model ?? "nbe";
  if (!SUPPORTED_MODELS.has(model)) {
    throw new Error(
      `iemMosForecasts: model must be one of ${[...SUPPORTED_MODELS].sort().join(",")}; got ${model}`,
    );
  }
  const fetchFn = opts.fetchFn ?? fetch;
  const fromDt = parseIsoDate(fromDate, false);
  const toDt = parseIsoDate(toDate, true);
  const hours = runtimeHoursFor(model, fromDt, toDt);
  const retrievedAt = new Date().toISOString();

  // GH #58: collect every (day × runtime-hour) URL first, then fan out
  // concurrently with a bounded pool. The cycles are independent (nothing in
  // one depends on another's response), so the previous serial-await loop
  // was paying N round-trips of ~620–760 ms each (~3.2 s for 12 cycles).
  // Bounded fan-out (MOS_FETCH_CONCURRENCY) collapses the wall-clock to
  // ~ceil(N/limit) round-trips on typical short windows while staying polite
  // on large historical ranges (an unbounded Promise.all over a year-scale
  // window would launch ~1,460 simultaneous requests — codex review P2).
  // Row ordering is preserved by keeping the URL list in day-then-hour order
  // and indexing results by position — byte-identical output to the serial path.
  const dayMs = 86_400_000;
  const urls: string[] = [];
  for (let day = fromDt.getTime(); day <= toDt.getTime(); day += dayMs) {
    for (const h of hours) {
      const rt = new Date(day);
      rt.setUTCHours(h, 0, 0, 0);
      if (rt < fromDt || rt > toDt) continue;
      // IEM /api/1/mos.json regex ^(AVN|GFS|...|NBE|...)$ is uppercase-only;
      // sending lowercase returns HTTP 422 (issue #17). Mirrors the upper()
      // applied to `model` on returned rows above and the Python fix at
      // _iem_mos.py:239.
      urls.push(
        `${IEM_MOS_URL}?station=${encodeURIComponent(
          station,
        )}&model=${encodeURIComponent(model.toUpperCase())}&runtime=${encodeURIComponent(
          rt.toISOString(),
        )}`,
      );
    }
  }

  // Run the FULL per-cycle lifecycle (fetch → status check → body read → row
  // projection) inside the bounded pool, NOT just the header fetch. `fetch()`
  // resolves once headers arrive, so bounding only the fetch promise would
  // still leave unbounded response bodies open and defer error handling until
  // every URL had been requested (codex review iter-2 P2). Returning per-cycle
  // row arrays and flattening in input order keeps byte-identical output.
  const perCycle = await mapWithConcurrency(urls, MOS_FETCH_CONCURRENCY, async (url) => {
    const resp = (await fetchFn(url)) as Response;
    if (resp.status === 404) return [] as IemMosRow[];
    if (!resp.ok) {
      throw new Error(`iemMosForecasts: HTTP ${resp.status} on ${url}`);
    }
    const payload = (await resp.json()) as { data?: RawMosRow[] };
    const out: IemMosRow[] = [];
    for (const raw of payload.data ?? []) {
      const projected = parseRow(raw, station, model, retrievedAt);
      if (projected !== null) out.push(projected);
    }
    return out;
  });
  return perCycle.flat();
}

export const __internal__ = {
  runtimeHoursFor,
  parseRow,
  NBE_CYCLE_CUTOVER,
};
